{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/srcastillo/NGS-Analyzer/blob/master/TALE_Writer_Colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "w52AyQ44SbAS"
      },
      "source": [
        "# TALE Writer Colab\n",
        "\n",
        "This Colab notebook allows you to design TALE nucleases (TALENs), and FusX TALE Base Editors (FusXTBEs) for C-to-T editing. TALENs can be designed against a specific target coordinate or against multiple targets within a given sequence. FusXTBEs can be designed against all 5'-TC-3' target sites within a given sequence or for premature termination codon (PTC) induction in a protein-coding gene. Subsequently, the repeat variable diresidues (RVDs) of the left and right TALE binding sequences and their corresponding FusX recipes can be obtained.\n",
        "\n",
        "**Inputs & outputs**\n",
        "\n",
        "A detailed explanation of all inputs and outputs is available in the [TALE Writer README](https://github.com/srcastillo/TALE-Writer/blob/main/README.md#inputs--outputs).\n",
        "\n",
        "**General information**\n",
        "\n",
        "* There can be dozens of design permutations (pairs of TALE binding sequences) per target. One way to decrease output size is to define constrained custom design parameters. It is up to the user to select the design permutation that best fits their specific application.\n",
        "\n",
        "* The [FusX recipe tool](http://www.talendesign.org/pFUXrecipeInput.php) referenced in this script works for TALE arrays targeting sequences 15- to 17-bp-long. To assemble TALE-based technologies with TALE arrays that target shorter or longer sequences, alternative assembly approaches or adaper kits must be used.\n",
        "\n",
        "* Loss and gain of restriction sites for genotyping:\n",
        "  * In TALEN design, TALE Writer identifies potential loss of restriction sites within the target region.\n",
        "  * In FusXTBE design, TALE Writer identifies both loss and gain of restriction sites within the target region.\n",
        "\n",
        "* For PTC induction with FusXTBEs, TALE Writer identifies the following sequence motifs:\n",
        "\n",
        "  *   The 5’-TGA-3’ motif (TGA in -frame), in which a G-to-A edit results in the stop codon UAA.\n",
        "  *   The 5’-TCAA-3’ motif (CAA in-frame), in which a C-to-T edit results in the stop codon UAA.\n",
        "  *   The 5’-TCAG-3’ motif (CAG in-frame), in which a C-to-T edit results in the stop codon UAG.\n",
        "\n",
        "* You can find more information about TALENs, mitochondrial C-to-T base editors, the FusX assembly system, and TALE Writer here:\n",
        "\n",
        "  * [CRISPR-free mitochondrial base editing paper](https://www.nature.com/articles/s41586-020-2477-4)\n",
        "  * [FusX TALE base editor (FusXTBE) paper](https://www.liebertpub.com/doi/full/10.1089/crispr.2021.0061)\n",
        "  * [FusX assembly system paper](https://www.liebertpub.com/doi/10.1089/hum.2015.172?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed)\n",
        "  * [TALE Writer on GitHub](https://github.com/srcastillo/TALE-Writer)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup\n",
        "Start by running the 2 cells below to set up TALE Writer and all required software."
      ],
      "metadata": {
        "id": "TPPECGYO_d9x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 1. Install third-party software\n",
        "\n",
        "#@markdown Please execute this cell by pressing the _Play_ button on the left to download and import third-party software in this Colab notebook.\n",
        "\n",
        "#@markdown **Note**: This installs the software on the Colab notebook in the cloud and not on your computer.\n",
        "\n",
        "# Local installations\n",
        "!pip install colorama\n",
        "!pip install tabulate"
      ],
      "metadata": {
        "cellView": "form",
        "id": "2JolxOV8BKx1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "scrolled": true,
        "id": "2GjrNfmdSbAY",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 2. Define libraries and functions\n",
        "\n",
        "#@markdown Please execute this cell by pressing the *Play* button on the left.\n",
        "\n",
        "# LIBRARIES, FUNCTIONS, AND INITIALIZATIONS\n",
        "\n",
        "# Libraries\n",
        "from colorama import Back, Style\n",
        "import matplotlib.pyplot as plt\n",
        "from tabulate import tabulate\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import sys\n",
        "\n",
        "# Find reverse complement\n",
        "def reverse_complement(sequence):\n",
        "    sequence, newSequence = sequence.lower(), ''\n",
        "    for i in range(0,len(sequence)):\n",
        "        # Unambiguous\n",
        "        if sequence[i] == 'a': newSequence = newSequence + 't'\n",
        "        elif sequence[i] == 'c': newSequence = newSequence + 'g'\n",
        "        elif sequence[i] == 'g': newSequence = newSequence + 'c'\n",
        "        elif sequence[i] == 't': newSequence = newSequence + 'a'\n",
        "        # Ambiguous\n",
        "        elif sequence[i] == 'r': newSequence = newSequence + 'y'\n",
        "        elif sequence[i] == 'y': newSequence = newSequence + 'r'\n",
        "        elif sequence[i] == 'k': newSequence = newSequence + 'm'\n",
        "        elif sequence[i] == 'm': newSequence = newSequence + 'k'\n",
        "        elif sequence[i] == 's': newSequence = newSequence + 's'\n",
        "        elif sequence[i] == 'w': newSequence = newSequence + 'w'\n",
        "        elif sequence[i] == 'b': newSequence = newSequence + 'v'\n",
        "        elif sequence[i] == 'd': newSequence = newSequence + 'h'\n",
        "        elif sequence[i] == 'h': newSequence = newSequence + 'd'\n",
        "        elif sequence[i] == 'v': newSequence = newSequence + 'b'\n",
        "        elif sequence[i] == 'n': newSequence = newSequence + 'n'\n",
        "    newSequence = newSequence[::-1]\n",
        "    return newSequence\n",
        "\n",
        "# Disambiguate DNA sequence\n",
        "def disambiguate(sequence):\n",
        "    sequence = sequence.lower()\n",
        "    # Calculate number of variations\n",
        "    ambiguous_list = ['R','Y','K','M','S','W','B','D','H','V','N']\n",
        "    possibilities_list = [2,2,2,2,2,2,3,3,3,3,4]\n",
        "    times_list, variations = [], 1\n",
        "    for i in range(0,len(ambiguous_list)):\n",
        "        times_list.append(sequence.count(ambiguous_list[i].lower()))\n",
        "        if times_list[i] != 0: variations = variations*possibilities_list[i]**times_list[i]\n",
        "    # Calculate all variations (disambiguate)\n",
        "    interrogate_list, i, stop1 = [], 0, False\n",
        "    interrogate_list.append(sequence)\n",
        "    while stop1 == False:\n",
        "        stop2, j = False, 0\n",
        "        while stop2 == False:\n",
        "            # Double-check that disambiguation is required\n",
        "            c = 0\n",
        "            for k in range(0,len(ambiguous_list)):\n",
        "                if (interrogate_list[i].find(ambiguous_list[k].lower()) == -1) == False: c += 1\n",
        "            if c == 0: stop1, stop2 = True, True\n",
        "            # Disambiguate symbol-by-symbol\n",
        "            if stop1 == False:\n",
        "                if interrogate_list[i][j] not in {'a','c','g','t'}: # If symbol is ambiguous\n",
        "                    stop2 = True    \n",
        "                    if interrogate_list[i][j] == 'r': # Disambiguate symbol\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'y':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'k':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'm':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 's':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'w':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'b':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'd':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'h':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'v':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                    elif interrogate_list[i][j] == 'n':\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'a' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'c' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 'g' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                        interrogate_list.append(interrogate_list[i][0:j] + 't' + interrogate_list[i][j+1:len(interrogate_list[i])])\n",
        "                elif interrogate_list[i][j] in {'a','c','g','t'}: # If symbol is unambiguous\n",
        "                    j += 1  \n",
        "        i += 1 \n",
        "    # Get rid of ambiguous seeds\n",
        "    variations_list = []\n",
        "    for i in range(len(interrogate_list)-variations,len(interrogate_list)): \n",
        "        variations_list.append(interrogate_list[i])\n",
        "    return variations_list\n",
        "\n",
        "# Align two sequences ('recognition' must be in 'sequence')\n",
        "def align(sequence, recognition):\n",
        "    sequence, recognition = sequence.lower(), recognition.lower()\n",
        "    x, y = 0, len(recognition)\n",
        "    i, j = 0, 0\n",
        "    while x < y:\n",
        "        if sequence[i] != recognition[j]:\n",
        "            recognition = ' ' + recognition\n",
        "            i, j = i + 1, j + 1\n",
        "            for k in range(0, x): \n",
        "                i, j = i - 1, j - 1\n",
        "            x = 0\n",
        "        elif sequence[i] == recognition[j]: \n",
        "            i, j, x = i + 1, j + 1, x + 1\n",
        "    z = len(recognition) - len(recognition.lstrip())\n",
        "    return [recognition, z]\n",
        "\n",
        "# Enzymes and their recognition sequences\n",
        "enzymes_list = ['AatII','Acc65I','AccI','AciI','AclI','AfeI','AflII','AflIII','AgeI','AhdI','AleI','AluI','AlwNI','ApaI','ApaLI','ApeKI','ApoI','AscI','AseI','AsiSI','AvaI','AvaII','AvrII','BaeGI','BamHI','BanI','BanII','BbvCI','BclI','BfaI','BglI','BglII','BlpI','BmgBI','BmtI','Bpu10I','BsaAI','BsaBI','BsaHI','BsaJI','BsaWI','BseYI','BsiEI','BsiHKAI','BsiWI','BslI','BsoBI','Bsp1286I','BspDI','BspEI','BspHI','BsrBI','BsrFI','BsrGI','BssHII','BssSI','BstAPI','BstBI','BstEII','BstNI','BstUI','BstXI','BstYI','BstZ17I','Bsu36I','BtgI','Cac8I','ClaI','CviAII','CviKI-1','CviQI','DdeI','DpnII','DraI','DraIII','DrdI','EaeI','EagI','Eco53kI','EcoNI','EcoO109I','EcoRI','EcoRV','FatI','Fnu4HI','FseI','FspI','HaeII','HaeIII','HhaI','HinP1I','HincII','HindIII','HinfI','HpaI','HpaII','Hpy166II','Hpy188I','Hpy188III','Hpy99I','HpyCH4III','HpyCH4IV','HpyCH4V','I-CeuI','I-SceI','KasI','KpnI','MboI','MfeI','MluCI','MluI','MscI','MseI','MslI','MspA1I','MspI','MwoI','NaeI','NarI','NciI','NcoI','NdeI','NgoMIV','NheI','NlaIII','NlaIV','NotI','NruI','NsiI','NspI','PI-PspI','PI-SceI','PacI','PaeR7I','PciI','PflFI','PflMI','PluTI','PmeI','PmlI','PpuMI','PshAI','PsiI','PspGI','PspOMI','PspXI','PstI','PvuI','PvuII','RsaI','RsrII','SacI','SacII','SalI','Sau3AI','Sau96I','SbfI','ScaI','ScrFI','SexAI','SfcI','SfiI','SfoI','SgrAI','SmaI','SmlI','SnaBI','SpeI','SphI','SrfI','SspI','StuI','StyD4I','StyI','SwaI','TaqI','TfiI','TseI','Tsp45I','TspMI','TspRI','Tth111I','XbaI','XcmI','XhoI','XmaI','XmnI','ZraI','AcuI','AlwI','BaeI','BbsI','BbvI','BccI','BceAI','BcgI','BciVI','BcoDI','BfuAI','BmrI','BpmI','BpuEI','BsaI','BsaXI','BseRI','BsgI','BsmAI','BsmBI','BsmFI','BsmI','BspCNI','BspMI','BspQI','BsrDI','BsrI','BtgZI','BtsCI','BtsI','BtsIMutI','CspCI','EarI','EciI','Esp3I','FauI','FokI','HgaI','HphI','HpyAV','MboII','MlyI','MmeI','MnlI','NmeAIII','PaqCI','PleI','SapI','SfaNI']\n",
        "recognition_list, complements_list = ['GACGTC','GGTACC','GTMKAC','CCGC','AACGTT','AGCGCT','CTTAAG','ACRYGT','ACCGGT','GACNNNNNGTC','CACNNNNGTG','AGCT','CAGNNNCTG','GGGCCC','GTGCAC','GCWGC','RAATTY','GGCGCGCC','ATTAAT','GCGATCGC','CYCGRG','GGWCC','CCTAGG','GKGCMC','GGATCC','GGYRCC','GRGCYC','CCTCAGC','TGATCA','CTAG','GCCNNNNNGGC','AGATCT','GCTNAGC','CACGTC','GCTAGC','CCTNAGC','YACGTR','GATNNNNATC','GRCGYC','CCNNGG','WCCGGW','CCCAGC','CGRYCG','GWGCWC','CGTACG','CCNNNNNNNGG','CYCGRG','GDGCHC','ATCGAT','TCCGGA','TCATGA','CCGCTC','RCCGGY','TGTACA','GCGCGC','CACGAG','GCANNNNNTGC','TTCGAA','GGTNACC','CCWGG','CGCG','CCANNNNNNTGG','RGATCY','GTATAC','CCTNAGG','CCRYGG','GCNNGC','ATCGAT','CATG','RGCY','GTAC','CTNAG','GATC','TTTAAA','CACNNNGTG','GACNNNNNNGTC','YGGCCR','CGGCCG','GAGCTC','CCTNNNNNAGG','RGGNCCY','GAATTC','GATATC','CATG','GCNGC','GGCCGGCC','TGCGCA','RGCGCY','GGCC','GCGC','GCGC','GTYRAC','AAGCTT','GANTC','GTTAAC','CCGG','GTNNAC','TCNGA','TCNNGA','CGWCG','ACNGT','ACGT','TGCA','TAACTATAACGGTCCTAAGGTAGCGAA','TAGGGATAACAGGGTAAT','GGCGCC','GGTACC','GATC','CAATTG','AATT','ACGCGT','TGGCCA','TTAA','CAYNNNNRTG','CMGCKG','CCGG','GCNNNNNNNGC','GCCGGC','GGCGCC','CCSGG','CCATGG','CATATG','GCCGGC','GCTAGC','CATG','GGNNCC','GCGGCCGC','TCGCGA','ATGCAT','RCATGY','TGGCAAACAGCTATTATGGGTATTATGGGT','ATCTATGTCGGGTGCGGAGAAAGAGGTAATGAAATGG','TTAATTAA','CTCGAG','ACATGT','GACNNNGTC','CCANNNNNTGG','GGCGCC','GTTTAAAC','CACGTG','RGGWCCY','GACNNNNGTC','TTATAA','CCWGG','GGGCCC','VCTCGAGB','CTGCAG','CGATCG','CAGCTG','GTAC','CGGWCCG','GAGCTC','CCGCGG','GTCGAC','GATC','GGNCC','CCTGCAGG','AGTACT','CCNGG','ACCWGGT','CTRYAG','GGCCNNNNNGGCC','GGCGCC','CRCCGGYG','CCCGGG','CTYRAG','TACGTA','ACTAGT','GCATGC','GCCCGGGC','AATATT','AGGCCT','CCNGG','CCWWGG','ATTTAAAT','TCGA','GAWTC','GCWGC','GTSAC','CCCGGG','CASTG','GACNNNGTC','TCTAGA','CCANNNNNNNNNTGG','CTCGAG','CCCGGG','GAANNNNTTC','GACGTC','CTGAAG','GGATC','ACNNNNGTAYC','GAAGAC','GCAGC','CCATC','ACGGC','CGANNNNNNTGC','GTATCC','GTCTC','ACCTGC','ACTGGG','CTGGAG','CTTGAG','GGTCTC','ACNNNNNCTCC','GAGGAG','GTGCAG','GTCTC','CGTCTC','GGGAC','GAATGC','CTCAG','ACCTGC','GCTCTTC','GCAATG','ACTGG','GCGATG','GGATG','GCAGTG','CAGTG','CAANNNNNGTGG','CTCTTC','GGCGGA','CGTCTC','CCCGC','GGATG','GACGC','GGTGA','CCTTC','GAAGA','GAGTC','TCCRAC','CCTC','GCCGAG','CACCTGC','GAGTC','GCTCTTC','GCATC'], []\n",
        "for i in range(0,len(recognition_list)): \n",
        "    recognition_list[i] = recognition_list[i].lower()\n",
        "    complements_list.append(reverse_complement((recognition_list[i])))\n",
        "# Enzymes with unambiguous recognition sequences\n",
        "fullEnzymes_list, fullRecognition_list, fullComplements_list = [], [], []\n",
        "for i in range(0,len(recognition_list)):\n",
        "    x = len(disambiguate(recognition_list[i]))\n",
        "    if x == 1:\n",
        "        fullEnzymes_list.append(enzymes_list[i])\n",
        "        fullRecognition_list.append(recognition_list[i])\n",
        "        fullComplements_list.append(complements_list[i])\n",
        "    elif x > 1:\n",
        "        if x <= 256: # Cutoff to decrease computation time\n",
        "            if recognition_list[i] == complements_list[i]:\n",
        "                for j in range(0,x):\n",
        "                    fullEnzymes_list.append(enzymes_list[i])\n",
        "                    y = disambiguate(recognition_list[i])[j]\n",
        "                    fullRecognition_list.append(y)\n",
        "                    fullComplements_list.append(y)\n",
        "            elif recognition_list[i] != complements_list[i]:\n",
        "                for j in range(0,x):\n",
        "                    fullEnzymes_list.append(enzymes_list[i])\n",
        "                    fullRecognition_list.append(disambiguate(recognition_list[i])[j])\n",
        "                    fullComplements_list.append(disambiguate(complements_list[i])[j])\n",
        "print('Ready')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Design\n",
        "In this section, you will specify which TALE-based technology you wish to design, assign the user-defined parameters, and obtain a list of all the viable design permutations for your technology of interest. If you choose to save the output, the resulting CSV file will be available in the _Files_ folder to the left of the Colab interface.\n",
        "\n",
        "**Note:** All fields are validated to avoid errors.\n",
        "\n"
      ],
      "metadata": {
        "id": "nPy_zmm5_6Zd"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "scrolled": false,
        "id": "L3HFM-FASbAs",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 3. Find TALE binding sequences and download the results\n",
        "\n",
        "#@markdown Please execute this cell by pressing the _Play_ button on the left.\n",
        "\n",
        "# DESIGN PERMUTATIONS\n",
        "\n",
        "# Begin\n",
        "newAnalysis = 'yes'\n",
        "while newAnalysis == 'yes':\n",
        "\n",
        "    # TALENs or DdCBEs?\n",
        "    design = 0\n",
        "    while design == 0:\n",
        "        design = input('\\nDesign:\\nA. TALENs\\nB. DdCBEs (DddA)\\nC. DdCBEs (DddA11)\\n')\n",
        "        if design.lower() == 'exit': sys.exit(0)\n",
        "        if design not in {'A','a','B','b', 'C','c'}:\n",
        "            design = 0\n",
        "            print('Invalid input! Choose A, B or C.')\n",
        "        else: design = design.lower()  \n",
        "\n",
        "    # TALENs for what?\n",
        "    if design == 'a':\n",
        "        TALENfor = 0 \n",
        "        while TALENfor == 0:\n",
        "            TALENfor = input('\\nFind TALEN pairs for:\\nA. All potential targets\\nB. Specific site\\n')\n",
        "            if TALENfor.lower() == 'exit': sys.exit(0)\n",
        "            if TALENfor not in {'A','a','B','b'}:\n",
        "                TALENfor = 0\n",
        "                print('Invalid input! Choose A or B.')\n",
        "            else: TALENfor = TALENfor.lower()\n",
        "\n",
        "    # DdCBEs for what?\n",
        "    else:\n",
        "        BEfor = 0\n",
        "        while BEfor == 0:\n",
        "            BEfor = input(\"\\nBase editors for:\\nA. All potential targets\\nB. Specific site\\nC. PTCs\\n\")\n",
        "            if BEfor.lower() == 'exit': sys.exit(0)\n",
        "            if BEfor not in {'A','a','B','b','C','c'}:\n",
        "                BEfor = 0\n",
        "                print('Invalid input! Choose A, B or C.')\n",
        "            else: BEfor = BEfor.lower()\n",
        "\n",
        "    # Design parameters\n",
        "    default = 0\n",
        "    while default == 0:\n",
        "        default = input('\\nUse default design parameters? ')\n",
        "        if default.lower() == 'exit': sys.exit(0)\n",
        "        if default.lower() not in {'yes','y','no','n'}:\n",
        "            default = 0\n",
        "            print(\"Invalid input! Type 'yes' or 'no'\")     \n",
        "    # Default paramaters\n",
        "    if default.lower() in {'yes','y'}:\n",
        "        leftArmMin, spacerMin, rightArmMin, leftArm5 = 15,14,15, True\n",
        "        leftArmMax, spacerMax, rightArmMax, rightArm5 = 16,16,16, True\n",
        "        targetWindow = 4\n",
        "        print('\\nLeft arm minimum length (bp): 15\\nLeft arm maximum length (bp): 16\\n'\\\n",
        "              \"Left 5'-T constraint: Yes\\n\"\\\n",
        "              '\\nSpacer minimum length (bp): 14\\nSpacer maximum length (bp): 16\\n'\\\n",
        "              '\\nRight arm minimum length (bp): 15\\nRight arm maximum length (bp): 16\\n'\\\n",
        "              \"Right 5'-T constraint: Yes\\n\"\\\n",
        "              '\\nTarget window: 4')\n",
        "    # Custom parameters\n",
        "    else:\n",
        "        # Left arm minimum length\n",
        "        leftArmMin = 0\n",
        "        while leftArmMin == 0:\n",
        "            leftArmMin = input('\\nLeft arm minimum length (bp): ')\n",
        "            if leftArmMin.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                leftArmMin = int(leftArmMin)\n",
        "                if leftArmMin < 5 or leftArmMin > 25:\n",
        "                    leftArmMin = 0\n",
        "                    print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "            except:\n",
        "                leftArmMin = 0\n",
        "                print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "        # Left arm maximum length\n",
        "        leftArmMax = 0\n",
        "        while leftArmMax == 0:\n",
        "            leftArmMax = input('\\nLeft arm maximum length (bp): ')\n",
        "            if leftArmMax.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                leftArmMax = int(leftArmMax)\n",
        "                if leftArmMax < 5 or leftArmMax > 25:\n",
        "                    leftArmMax = 0\n",
        "                    print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "                elif leftArmMax < leftArmMin:\n",
        "                    leftArmMax = 0\n",
        "                    print('Invalid input! Maximum length cannot be less than minimum length.')\n",
        "            except:\n",
        "                leftArmMax = 0\n",
        "                print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "        # Left arm 5'-T constraint\n",
        "        leftArm5 = 2\n",
        "        while leftArm5 == 2:\n",
        "            leftArm5 = input(\"\\nLeft 5'-T constraint? \")\n",
        "            if leftArm5.lower() == 'exit': sys.exit(0)\n",
        "            if leftArm5.lower() not in {'yes','y','no','n'}:\n",
        "                leftArm5 = 2\n",
        "                print(\"Invalid input! Type 'yes' or 'no'\")\n",
        "            elif leftArm5.lower() in {'yes','y'}: leftArm5 = True\n",
        "            elif leftArm5.lower() in {'no','n'}: leftArm5 = False\n",
        "        # Spacer minimum length\n",
        "        spacerMin = 0\n",
        "        while spacerMin == 0:\n",
        "            spacerMin = input('\\nSpacer minimum length (bp): ')\n",
        "            if spacerMin.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                spacerMin = int(spacerMin)\n",
        "                if spacerMin < 5 or spacerMin > 25:\n",
        "                    spacerMin = 0\n",
        "                    print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "            except:\n",
        "                spacerMin = 0\n",
        "                print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "        # Spacer maximum length\n",
        "        spacerMax = 0\n",
        "        while spacerMax == 0:\n",
        "            spacerMax = input('\\nSpacer maximum length (bp): ')\n",
        "            if spacerMax.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                spacerMax = int(spacerMax)\n",
        "                if spacerMax < 5 or spacerMax > 25:\n",
        "                    spacerMax = 0\n",
        "                    print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "                elif spacerMax < spacerMin:\n",
        "                    spacerMax = 0\n",
        "                    print('Invalid input! Maximum length cannot be less than minimum length.')\n",
        "            except:\n",
        "                spacerMax = 0\n",
        "                print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "        # Right arm minimum length\n",
        "        rightArmMin = 0\n",
        "        while rightArmMin == 0:\n",
        "            rightArmMin = input('\\nRight arm minimum length (bp): ')\n",
        "            if rightArmMin.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                rightArmMin = int(rightArmMin)\n",
        "                if rightArmMin < 5 or rightArmMin > 25:\n",
        "                    rightArmMin = 0\n",
        "                    print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "            except:\n",
        "                rightArmMin = 0\n",
        "                print('Invalid input! Minimum length must be an integer between 5 and 25.')\n",
        "        # Right arm maximum length\n",
        "        rightArmMax = 0\n",
        "        while rightArmMax == 0:\n",
        "            rightArmMax = input('\\nRight arm maximum length (bp): ')\n",
        "            if rightArmMax.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                rightArmMax = int(rightArmMax)\n",
        "                if rightArmMax < 5 or rightArmMax > 25:\n",
        "                    rightArmMax = 0\n",
        "                    print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "                elif rightArmMax < rightArmMin:\n",
        "                    rightArmMax = 0\n",
        "                    print('Invalid input! Maximum length cannot be less than minimum length.')\n",
        "            except:\n",
        "                rightArmMax = 0\n",
        "                print('Invalid input! Maximum length must be an integer between 5 and 25.')\n",
        "        # Right arm 5'-T constraint\n",
        "        rightArm5 = 2\n",
        "        while rightArm5 == 2:\n",
        "            rightArm5 = input(\"\\nRight 5'-T constraint? \")\n",
        "            if rightArm5.lower() == 'exit': sys.exit(0)\n",
        "            if rightArm5.lower() not in {'yes','y','no','n'}:\n",
        "                rightArm5 = 2\n",
        "                print(\"Invalid input! Type 'yes' or 'no'\")\n",
        "            elif rightArm5.lower() in {'yes','y'}: rightArm5 = True\n",
        "            elif rightArm5.lower() in {'no','n'}: rightArm5 = False\n",
        "        # Target window\n",
        "        targetWindow = -1\n",
        "        while targetWindow == -1:\n",
        "            targetWindow = input('\\nTarget window: ')\n",
        "            if targetWindow.lower() == 'exit': sys.exit(0)\n",
        "            try:\n",
        "                targetWindow = int(targetWindow)\n",
        "                # If spacer length is even\n",
        "                if (spacerMax % 2) == 0:\n",
        "                    if targetWindow < 0 or targetWindow > ((spacerMax-2)/2):\n",
        "                        targetWindow = -1\n",
        "                        print('Invalid input! The target window cannot be less than 0 or greater than ' + str(int((spacerMax-2)/2)) + '.')\n",
        "                # If spacer length is odd\n",
        "                else:\n",
        "                    if targetWindow < 0 or targetWindow > ((spacerMax-1)/2):\n",
        "                        targetWindow = -1\n",
        "                        print('Invalid input! The target window cannot be less than 0 or greater than ' + str(int((spacerMax-1)/2)) + '.')\n",
        "            except:\n",
        "                targetWindow = -1\n",
        "                print('Invalid input! The target window must be a non-negative integer.')\n",
        "\n",
        "    # Begin design analysis\n",
        "    newSequence, count = True, 0\n",
        "    while newSequence == True:\n",
        "\n",
        "        # Analysis number\n",
        "        count += 1\n",
        "        print('')\n",
        "        print('-' * (13 + len(str(count))))\n",
        "        print('SEQUENCE NO. ' + str(count))\n",
        "        print('-' * (13 + len(str(count))))\n",
        "\n",
        "        # Define input\n",
        "        sequence = 0\n",
        "        while sequence == 0:\n",
        "            sequence = input(\"\\nSequence (5'-to-3'): \")\n",
        "            if sequence.lower() == 'exit': sys.exit(0)\n",
        "            x, i = sequence.lower(), 0\n",
        "            while i in range(0,int(len(x))):\n",
        "                if x[i] not in {'a','c','g','t'}:\n",
        "                    print('Invalid sequence!')\n",
        "                    i = len(sequence)\n",
        "                    sequence = 0\n",
        "                else: i += 1\n",
        "        sequence = sequence.lower() # To facilitate the analysis\n",
        "\n",
        "        # TALEN design\n",
        "        if design == 'a':\n",
        "            # All potential targets\n",
        "            if TALENfor == 'a':\n",
        "                coordinates = [[],[]]    \n",
        "                for i in range(0,len(sequence)-1):\n",
        "                    coordinates[0].append(sequence[i:i+2].upper())\n",
        "                    coordinates[1].append(i+1)   \n",
        "            # Specific site\n",
        "            elif TALENfor == 'b':\n",
        "                # Define the target\n",
        "                coordinate, coordinates = 0, [[],[]]\n",
        "                while coordinate == 0:\n",
        "                    coordinate = input('\\nTarget coordinate: ')\n",
        "                    if coordinate.lower() == 'exit': sys.exit(0)\n",
        "                    try:\n",
        "                        coordinate = int(coordinate)\n",
        "                        if coordinate <= 0:\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate must be a positive integer.')\n",
        "                        elif coordinate > len(sequence):\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate cannot be outside of the sequence of interest.')\n",
        "                        else:\n",
        "                            coordinates[0].append(sequence[coordinate-1:coordinate-1+2].upper())\n",
        "                            coordinates[1].append(coordinate)\n",
        "                    except ValueError:\n",
        "                        coordinate = 0\n",
        "                        print('Invalid input! The target coordinate must be a positive integer.')\n",
        "            proceed = True # No need to check for targets\n",
        "\n",
        "        # TC-to-TT DdCBE design\n",
        "        elif design == 'b':\n",
        "            coordinates, targetCount = [[],[]], 0\n",
        "            # TC motifs\n",
        "            if BEfor == 'a':\n",
        "                # Find all target sites\n",
        "                for i in range(0,int(len(sequence))):\n",
        "                    if sequence[i:i+2] == 'tc' or sequence[i:i+2] == 'ga':\n",
        "                        coordinates[0].append(sequence[i:i+2])\n",
        "                        coordinates[1].append(i+1)\n",
        "                        targetCount += 1                \n",
        "            # Specific site\n",
        "            elif BEfor == 'b':\n",
        "                # Define the target\n",
        "                coordinate = 0\n",
        "                while coordinate == 0:\n",
        "                    coordinate = input('\\nTarget coordinate: ')\n",
        "                    if coordinate.lower() == 'exit': sys.exit(0)\n",
        "                    try:\n",
        "                        coordinate = int(coordinate)\n",
        "                        if coordinate <= 0:\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate must be a positive integer.')\n",
        "                        elif coordinate > len(sequence):\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate cannot be outside of the sequence of interest.')\n",
        "                        else:\n",
        "                            coordinates[0].append(sequence[coordinate-1:coordinate-1+2].upper())\n",
        "                            coordinates[1].append(coordinate)\n",
        "                    except ValueError:\n",
        "                        coordinate = 0\n",
        "                        print('Invalid input! The target coordinate must be a positive integer.')\n",
        "                # Check for a target in that specific site\n",
        "                x = sequence[coordinate-1:coordinate+1]\n",
        "                if x == 'tc' or x == 'ga': targetCount += 1\n",
        "            # PTC motifs\n",
        "            elif BEfor == 'c':\n",
        "                triplets = []\n",
        "                for i in range(0,int(len(sequence)/3)):\n",
        "                    triplets.append(sequence[3*i:3*i+3])\n",
        "                    # UAA\n",
        "                    if triplets[i] == 'tga': # TGA -> TAA -> UAA\n",
        "                        coordinates[0].append('TGA')\n",
        "                        coordinates[1].append(3*i+1)\n",
        "                        targetCount += 1\n",
        "                    elif triplets[i] == 'caa': # T|CAA -> T|TAA -> U|UAA\n",
        "                        if i != 0:\n",
        "                            if triplets[i-1][2] == 't':\n",
        "                                coordinates[0].append('TCAA')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                    # UAG\n",
        "                    elif triplets[i] == 'cag': # T|CAG -> T|TAG -> U|UAG\n",
        "                        if i != 0:\n",
        "                            if triplets[i-1][2] == 't':\n",
        "                                coordinates[0].append('TCAG')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1   \n",
        "            proceed = targetCount != 0                    \n",
        "        \n",
        "        # HC-to-HT DdCBE design\n",
        "        elif design == 'c':\n",
        "            coordinates, targetCount = [[],[]], 0\n",
        "            # HC motifs\n",
        "            if BEfor == 'a':\n",
        "                # Find all target sites\n",
        "                for i in range(0,int(len(sequence))):\n",
        "                    # 5'-TC-3' or 5'-GA-3' motifs\n",
        "                    if sequence[i:i+2] == 'tc' or sequence[i:i+2] == 'ga':\n",
        "                        coordinates[0].append(sequence[i:i+2])\n",
        "                        coordinates[1].append(i+1)\n",
        "                        targetCount += 1\n",
        "                    # 5'-CC-3' or 5'-GG-3' motifs\n",
        "                    elif sequence[i:i+2] == 'cc' or sequence[i:i+2] == 'gg':\n",
        "                        coordinates[0].append(sequence[i:i+2])\n",
        "                        coordinates[1].append(i+1)\n",
        "                        targetCount += 1\n",
        "                    # 5'-AC-3' or 5'-GT-3' motifs\n",
        "                    elif sequence[i:i+2] == 'ac' or sequence[i:i+2] == 'gt':\n",
        "                        coordinates[0].append(sequence[i:i+2])\n",
        "                        coordinates[1].append(i+1)\n",
        "                        targetCount += 1\n",
        "            # Specific site\n",
        "            elif BEfor == 'b':\n",
        "                # Define the target\n",
        "                coordinate = 0\n",
        "                while coordinate == 0:\n",
        "                    coordinate = input('\\nTarget coordinate: ')\n",
        "                    if coordinate.lower() == 'exit': sys.exit(0)\n",
        "                    try:\n",
        "                        coordinate = int(coordinate)\n",
        "                        if coordinate <= 0:\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate must be a positive integer.')\n",
        "                        elif coordinate > len(sequence):\n",
        "                            coordinate = 0\n",
        "                            print('Invalid input! The target coordinate cannot be outside of the sequence of interest.')\n",
        "                        else:\n",
        "                            coordinates[0].append(sequence[coordinate-1:coordinate-1+2].upper())\n",
        "                            coordinates[1].append(coordinate)\n",
        "                    except ValueError:\n",
        "                        coordinate = 0\n",
        "                        print('Invalid input! The target coordinate must be a positive integer.')  \n",
        "                # Check for a target in that specific site\n",
        "                x = sequence[coordinate-1:coordinate+1]\n",
        "                if x == 'tc' or x == 'cc' or x == 'ac' or x == 'ga' or x == 'gg' or x == 'gt': targetCount += 1\n",
        "            # PTC motifs\n",
        "            elif BEfor == 'c':\n",
        "                triplets = []\n",
        "                for i in range(0,int(len(sequence)/3)):\n",
        "                    triplets.append(sequence[3*i:3*i+3])\n",
        "                    # UAA\n",
        "                    if triplets[i] == 'tga': # TGA -> TAA -> UAA\n",
        "                        coordinates[0].append('TGA')\n",
        "                        coordinates[1].append(3*i+1)\n",
        "                        targetCount += 1\n",
        "                    elif triplets[i] == 'caa': # H|CAA -> H|TAA -> H|UAA\n",
        "                        if i != 0:\n",
        "                            if triplets[i-1][2] == 't':\n",
        "                                coordinates[0].append('TCAA')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                            elif triplets[i-1][2] == 'c':\n",
        "                                coordinates[0].append('CCAA')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                            elif triplets[i-1][2] == 'a':\n",
        "                                coordinates[0].append('ACAA')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                    # UAG\n",
        "                    elif triplets[i] == 'tgg': # TGG -> TAG -> UAG\n",
        "                        coordinates[0].append('TGG')\n",
        "                        coordinates[1].append(3*i+1)\n",
        "                        targetCount += 1\n",
        "                    elif triplets[i] == 'cag': # H|CAG -> H|TAG -> H|UAG\n",
        "                        if i != 0:\n",
        "                            if triplets[i-1][2] == 't':\n",
        "                                coordinates[0].append('TCAG')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1 \n",
        "                            elif triplets[i-1][2] == 'c':\n",
        "                                coordinates[0].append('CCAG')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                            elif triplets[i-1][2] == 'a':\n",
        "                                coordinates[0].append('ACAG')\n",
        "                                coordinates[1].append(3*i)\n",
        "                                targetCount += 1\n",
        "                    # AGA\n",
        "                    elif triplets[i] == 'gga': # GGA -> AGA -> AGA\n",
        "                        coordinates[0].append('GGA')\n",
        "                        coordinates[1].append(3*i+1)\n",
        "                        targetCount += 1\n",
        "                    # AGG\n",
        "                    elif triplets[i] == 'ggg': # GGG -> AGG -> AGG\n",
        "                        coordinates[0].append('GGG')\n",
        "                        coordinates[1].append(3*i+1)\n",
        "                        targetCount += 1\n",
        "            proceed = targetCount != 0  \n",
        "        \n",
        "        # If there are motifs\n",
        "        if proceed:\n",
        "            # Identify all possible design permutations\n",
        "            design_list, RE_list = [[],[],[],[],[],[],[],[]], [[],[]]\n",
        "            for i in range(0,len(coordinates[0])): # Include all targets\n",
        "                designCounter = 0 # To count the number of designs\n",
        "                # Additional conditions to avoid 'index out of range' errors\n",
        "                condition5end = coordinates[1][i] >= leftArmMax + spacerMax\n",
        "                condition3end = len(sequence) - coordinates[1][i] >= spacerMax + rightArmMax\n",
        "                if condition5end and condition3end:  \n",
        "                    for j in range (spacerMin,spacerMax+1):\n",
        "                        for k in range(leftArmMin,leftArmMax+1):\n",
        "                            for l in range(rightArmMin,rightArmMax+1):\n",
        "                                for m in range(0,j-len(coordinates[0][i])+1): # Slide target through spacer\n",
        "                                    # Left arm\n",
        "                                    leftArm = sequence[coordinates[1][i]-2-m-k:coordinates[1][i]-1-m]\n",
        "                                    # Spacer (capitalize the targetable motif)\n",
        "                                    spacerX = sequence[coordinates[1][i]-1-m:coordinates[1][i]-1+j-m]\n",
        "                                    spacer = spacerX[0:m] + spacerX[m:m+len(coordinates[0][i])].upper() + spacerX[m+len(coordinates[0][i]):len(spacerX)]\n",
        "                                    # Right arm (matches the actual TALE binding sequence)\n",
        "                                    rightArm = sequence[coordinates[1][i]-1+j-m:coordinates[1][i]+j-m+l]\n",
        "                                    rightArm = reverse_complement(rightArm)\n",
        "                                    # Check the 5'-T constraint\n",
        "                                    if leftArm5 and rightArm5: constraint = leftArm[0] == 't' and rightArm[0] == 't'\n",
        "                                    elif not leftArm5 and rightArm5: constraint = rightArm[0] == 't'\n",
        "                                    elif leftArm5 and not rightArm5: constraint = leftArm[0] == 't'\n",
        "                                    elif not leftArm5 and not rightArm5: constraint = True\n",
        "                                    if constraint:\n",
        "                                        # The motif must be within the 'editing window'\n",
        "                                        if m+1 >= targetWindow and m+2 <= len(spacer) - targetWindow: \n",
        "                                            designCounter += 1 # Increase the design counter\n",
        "                                            # Append target coordinate, target motif, and design index\n",
        "                                            design_list[1].append(coordinates[1][i])\n",
        "                                            design_list[2].append(coordinates[0][i])\n",
        "                                            design_list[3].append(designCounter)\n",
        "                                            # Append off-target(s)\n",
        "                                            if design == 'a': design_list[4].append(0) # TALENs\n",
        "                                            else: # DdCBEs\n",
        "                                                offTargetsTC = spacer.count('tc') + spacer.count('ga')\n",
        "                                                # TC-to-TT\n",
        "                                                if design == 'b': \n",
        "                                                    design_list[4].append(offTargetsTC) \n",
        "                                                # HC-to-HT\n",
        "                                                elif design == 'c': \n",
        "                                                    offTargetsCC = spacer.count('cc') + spacer.count('gg')\n",
        "                                                    offTargetsAC = spacer.count('ac') + spacer.count('gt')\n",
        "                                                    design_list[4].append(offTargetsTC + offTargetsCC + offTargetsAC)\n",
        "                                            # Append left and right arms\n",
        "                                            design_list[5].append(leftArm[1:len(leftArm)]) \n",
        "                                            design_list[6].append(rightArm[1:len(rightArm)])\n",
        "                                            # Identify potential loss of restriction sites in the spacer\n",
        "                                            alignmentStarts, alignmentEnds, REc, REs = 0, 0, 0, ''\n",
        "                                            for n in range(0,len(fullEnzymes_list)):\n",
        "                                                # Take advantage of palindromy to decrease computation time\n",
        "                                                if fullRecognition_list[n] == fullComplements_list[n]:\n",
        "                                                    # Check if the recognition sequence is in the spacer\n",
        "                                                    if fullRecognition_list[n] in spacer.lower():\n",
        "                                                        # Define the limits of the valid alignment range\n",
        "                                                        alignmentStarts = align(spacer, fullRecognition_list[n])[1] - 1\n",
        "                                                        alignmentEnds = len(align(spacer, fullRecognition_list[n])[0])\n",
        "                                                        # Check if the recognition sequence and the target overlap\n",
        "                                                        if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                            REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                                # If the restriction site isn't palindromic\n",
        "                                                elif fullRecognition_list[n] != fullComplements_list[n]:\n",
        "                                                    # Check that the recognition site is in the spacer\n",
        "                                                    if fullRecognition_list[n] in spacer.lower():\n",
        "                                                        # Define the limits of the valid alignment range \n",
        "                                                        alignmentStarts = align(spacer, fullRecognition_list[n])[1] - 1\n",
        "                                                        alignmentEnds = len(align(spacer, fullRecognition_list[n])[0])\n",
        "                                                        # Check if the recognition sequence and the target overlap\n",
        "                                                        if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                            REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                                    # Check if the reverse complement is in the spacer\n",
        "                                                    elif fullComplements_list[n] in spacer.lower():\n",
        "                                                        # Define the limits of the valid alignment range\n",
        "                                                        alignmentStarts = align(spacer, fullComplements_list[n])[1] - 1\n",
        "                                                        alignmentEnds = len(align(spacer, fullComplements_list[n])[0])\n",
        "                                                        # Check if the recognition sequence and the target overlap\n",
        "                                                        if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                            REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                            # Finishing touches\n",
        "                                            if REc != 0: REs = REs[:-1]; REs = REs[:-1]\n",
        "                                            elif REc == 0: REs = 'None'\n",
        "                                            RE_list[0].append(REs) # Append results\n",
        "                                            # Ambiguity (only deterministic in base editing)\n",
        "                                            if design != 'a':\n",
        "                                                # Identify restriction sites (potential gain)                                   \n",
        "                                                alignmentStarts, alignmentEnds, REc, REs = 0, 0, 0, ''\n",
        "                                                # Include on-target edits\n",
        "                                                spacerX = spacer\n",
        "                                                if 'TC' in spacerX: spacerX = spacerX.replace('C','T')\n",
        "                                                elif 'GA' in spacerX: spacerX = spacerX.replace('G','A')\n",
        "                                                if design == 'c': # Add CC-to-CT and AC-to-AT edits\n",
        "                                                    if 'CC' in spacerX: spacerX = spacerX[0:spacerX.index('C') + 1] + 'T' + spacerX[len(spacerX) - 1]\n",
        "                                                    elif 'GG' in spacerX: spacerX = spacerX[0:spacerX.index('G')] + 'C' + spacerX[spacerX.index('G')+1:len(spacerX)]\n",
        "                                                    elif 'AC' in spacerX: spacerX = spacerX.replace('C','T')\n",
        "                                                    elif 'GT' in spacerX: spacerX = spacerX.replace('G','A')\n",
        "                                                # Iterate over the enzymes and their recognition sequences\n",
        "                                                for n in range(0,len(fullEnzymes_list)):\n",
        "                                                    # Take advantage of palindromy to decrease computation time\n",
        "                                                    if fullRecognition_list[n] == fullComplements_list[n]:\n",
        "                                                        # Check if the recognition sequence is in the edited spacer                                        \n",
        "                                                        if fullRecognition_list[n] in spacerX.lower():\n",
        "                                                            # Define the limits of the valid alignment range\n",
        "                                                            alignmentStarts = align(spacerX, fullRecognition_list[n])[1] - 1\n",
        "                                                            alignmentEnds = len(align(spacerX, fullRecognition_list[n])[0])\n",
        "                                                            # Check if the recognition sequence and the target overlap\n",
        "                                                            if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                                REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                                    # If the restriction site isn't palindromic\n",
        "                                                    elif fullRecognition_list[n] != fullComplements_list[n]:\n",
        "                                                        # Check that the recognition site is in the edited spacer\n",
        "                                                        if fullRecognition_list[n] in spacerX.lower():\n",
        "                                                            # Define the limits of the valid alignment range \n",
        "                                                            alignmentStarts = align(spacerX, fullRecognition_list[n])[1] - 1\n",
        "                                                            alignmentEnds = len(align(spacerX, fullRecognition_list[n])[0])\n",
        "                                                            # Check if the recognition sequence and the target overlap\n",
        "                                                            if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                                REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                                        # Check if the reverse complement is in the edited spacer\n",
        "                                                        if fullComplements_list[n] in spacerX.lower():\n",
        "                                                            # Define the limits of the valid alignment range\n",
        "                                                            alignmentStarts = align(spacerX, fullComplements_list[n])[1] - 1\n",
        "                                                            alignmentEnds = len(align(spacerX, fullComplements_list[n])[0])\n",
        "                                                            # Check if the recognition sequence and the target overlap\n",
        "                                                            if m in range(alignmentStarts, alignmentEnds):\n",
        "                                                                REc, REs = REc + 1, REs + fullEnzymes_list[n] + ', '\n",
        "                                                # Finishing touches\n",
        "                                                if REc != 0: REs = REs[:-1]; REs = REs[:-1]\n",
        "                                                elif REc == 0: REs = 'None'\n",
        "                                                RE_list[1].append(REs) # Append results\n",
        "                                            # Append spacer (with highlights)\n",
        "                                            for n in {'A','C','G','T'}: spacer = spacer.replace(n, Back.BLUE + n + Style.RESET_ALL)\n",
        "                                            if design != 'a': # All DdCBEs\n",
        "                                                spacer = spacer.replace('tc', Back.RED + 'tc' + Style.RESET_ALL)\n",
        "                                                spacer = spacer.replace('ga', Back.RED + 'ga' + Style.RESET_ALL)\n",
        "                                                if design == 'c': # HC-to-TT DdCBEs\n",
        "                                                    spacer = spacer.replace('cc', Back.RED + 'cc' + Style.RESET_ALL)\n",
        "                                                    spacer = spacer.replace('gg', Back.RED + 'gg' + Style.RESET_ALL)\n",
        "                                                    spacer = spacer.replace('ac', Back.RED + 'ac' + Style.RESET_ALL)\n",
        "                                                    spacer = spacer.replace('gt', Back.RED + 'gt' + Style.RESET_ALL)\n",
        "                                            design_list[7].append(spacer)\n",
        "\n",
        "            # Remove deterministic ambiguity\n",
        "            if design == 'b': \n",
        "                # Remove false positives enzymes\n",
        "                for i in range(0,len(RE_list[0])):\n",
        "                    s_list = []\n",
        "                    # Find the false positives\n",
        "                    if RE_list[0][i] != 'None' and RE_list[1][i] != 'None':\n",
        "                        s0 = set(RE_list[0][i].replace(' ', '').split(','))\n",
        "                        s1 = set(RE_list[1][i].replace(' ', '').split(','))\n",
        "                        s_list = sorted(s0 & s1)\n",
        "                    # Remove them\n",
        "                    for j in range(0, len(s_list)):\n",
        "                        RE_list[0][i] = RE_list[0][i].replace(s_list[j], '')\n",
        "                        RE_list[1][i] = RE_list[1][i].replace(s_list[j], '')   \n",
        "                    # Finishing touches\n",
        "                    if s_list: \n",
        "                        for j in range(0,2):\n",
        "                            # Only commas and spaces\n",
        "                            p, q = RE_list[j][i] == '', RE_list[j][i] == ', '\n",
        "                            r = (RE_list[j][i]+RE_list[j][i]).find(RE_list[j][i],1,-1) != -1\n",
        "                            if p or q or r: RE_list[j][i] = 'None'\n",
        "                            # Commas, spaces, and letters\n",
        "                            if RE_list[j][i] != ', ':\n",
        "                                # Clean up the beginning if it starts with ','\n",
        "                                if RE_list[j][i][0] == ',':\n",
        "                                    while RE_list[j][i][0:2] == ', ':\n",
        "                                        RE_list[j][i] = RE_list[j][i][2:len(RE_list[j][i])] # HERE\n",
        "                                # Clean up the middle if it contains ', , '\n",
        "                                if ', , ' in RE_list[j][i]:\n",
        "                                    while ', , ' in RE_list[j][i]: RE_list[j][i] = RE_list[j][i].replace(', , ', ', ')                                    \n",
        "                                # Clean up the ending if it ends with ' '\n",
        "                                if RE_list[j][i][len(RE_list[j][i])-1] == ' ':\n",
        "                                    while RE_list[j][i][len(RE_list[j][i])-2:len(RE_list[j][i])] == ', ':\n",
        "                                        RE_list[j][i] = RE_list[j][i][0:len(RE_list[j][i])-2]\n",
        "\n",
        "            # Correctly append target numbers\n",
        "            targetCounter = 1\n",
        "            for i in range(0,len(design_list[1])):\n",
        "                if i > 0:\n",
        "                    if design_list[1][i] != design_list[1][i-1]:\n",
        "                        targetCounter += 1\n",
        "                design_list[0].append(targetCounter)\n",
        "\n",
        "            # If there are no viable design permutations\n",
        "            if not design_list[0]:\n",
        "                print('\\nThere are no viable design permutations!')\n",
        "                # Ask to try another sequence\n",
        "                newSequenceX = 0\n",
        "                while newSequenceX == 0:\n",
        "                    newSequenceX = input('\\nNew sequence? ')\n",
        "                    if newSequenceX.lower() == 'exit': sys.exit(0)\n",
        "                    newSequenceX = newSequenceX.lower()\n",
        "                    if newSequenceX not in {'yes','y','no','n'}:\n",
        "                        newSequenceX = 0\n",
        "                        print(\"Invalid input! Type 'yes' or 'no'\")\n",
        "                    elif newSequenceX in {'no','n'}: newSequence = False\n",
        "\n",
        "            # Highlight targetable motifs\n",
        "            elif design_list[0]:\n",
        "                targetLength = 0\n",
        "                for i in range(0,len(design_list[1])):\n",
        "                    targetLength = len(design_list[2][i])\n",
        "                    sequence = sequence[:design_list[1][i]-1]+sequence[design_list[1][i]-1:design_list[1][i]-1+targetLength].upper()+sequence[design_list[1][i]-1+targetLength:]           \n",
        "                # Add blue highlights to the entire sequence\n",
        "                for i in {'A','C','G','T'}: sequence = sequence.replace(i, Back.BLUE + i + Style.RESET_ALL)\n",
        "                print('\\nSequence (with target/s): ' + sequence + '\\n')\n",
        "\n",
        "                # DISPLAY RESULTS\n",
        "                \n",
        "                # Additional lists to avoid saturating the displayed df\n",
        "                loss_list, gain_list = [], []\n",
        "                for i in range(0,len(RE_list[0])):\n",
        "                    if RE_list[0][i] == 'None': loss_list.append('No')\n",
        "                    else: loss_list.append('Yes')\n",
        "                    if design != 'a': # DdCBEs\n",
        "                        if RE_list[1][i] == 'None': gain_list.append('No')\n",
        "                        else: gain_list.append('Yes')\n",
        "                \n",
        "                # Create a dataframe containing the results\n",
        "                if design == 'a': # TALENs\n",
        "                    if TALENfor == 'a': # All targets\n",
        "                        results_df = pd.DataFrame({'TGT':design_list[0],\n",
        "                                                   'COORD':design_list[1],\n",
        "                                                   'DSG':design_list[3],\n",
        "                                                   'LOSS':loss_list,\n",
        "                                                   'LEFT SEQUENCE':design_list[5],\n",
        "                                                   'RIGHT SEQUENCE':design_list[6],\n",
        "                                                   'SPACER SEQUENCE':design_list[7]})\n",
        "                    elif TALENfor == 'b': # Specific site\n",
        "                        results_df = pd.DataFrame({'DSG':design_list[3],\n",
        "                                                   'LOSS':loss_list,\n",
        "                                                   'LEFT SEQUENCE':design_list[5],\n",
        "                                                   'RIGHT SEQUENCE':design_list[6],\n",
        "                                                   'SPACER SEQUENCE':design_list[7]})\n",
        "                else: # DdCBEs\n",
        "                    results_df = pd.DataFrame({'TGT':design_list[0],\n",
        "                                               'COORD':design_list[1],\n",
        "                                               'DSG':design_list[3],\n",
        "                                               'LOSS':loss_list,\n",
        "                                               'GAIN':gain_list,\n",
        "                                               'OFF':design_list[4],\n",
        "                                               'LEFT SEQUENCE':design_list[5],\n",
        "                                               'RIGHT SEQUENCE':design_list[6],\n",
        "                                               'SPACER SEQUENCE':design_list[7]})\n",
        "                print(tabulate(results_df, showindex=False, headers=results_df.columns))\n",
        "\n",
        "                # Visualizing the design permutation distribution\n",
        "                if design_list[0][-1] > 1:\n",
        "                    print('\\nThere are ' + str(design_list[0][-1]) + ' potential targets and a total of ' + str(len(design_list[0])) + ' design permutations.')\n",
        "                    print('\\nDesign permutation distribution:')\n",
        "                    fig_size, fig_size[0], fig_size[1] = plt.rcParams[\"figure.figsize\"], 10, 6\n",
        "                    fig, ax = plt.subplots()\n",
        "                    ax.set(xlabel = 'Target', ylabel = 'Designs')\n",
        "                    permutationsGraph_list, subtract = [], 0\n",
        "                    for i in range(0,len(design_list[0])):\n",
        "                        if design_list[0][i] > design_list[0][i-1]:\n",
        "                            permutationsGraph_list.append(i-subtract)\n",
        "                            subtract = i       \n",
        "                    permutationsGraph_list.append(design_list[0].count(design_list[0][-1]))\n",
        "                    permutationsGraph = np.zeros((len(permutationsGraph_list)))\n",
        "                    for i in range(0,len(permutationsGraph_list)): permutationsGraph[i] = int(permutationsGraph_list[i])\n",
        "                    i = np.arange(1, design_list[0][-1] + 1, 1)\n",
        "                    ax.scatter(i, permutationsGraph, c = 'black', s = 4)\n",
        "                    ax.plot(i, permutationsGraph, c = 'gray', linewidth = 1, linestyle = '-') \n",
        "                    ax.set_ylim(0, np.amax(permutationsGraph) + 1)\n",
        "                    plt.show()\n",
        "                else: print('\\nThere is 1 potential target and ' + str(len(design_list[0])) + ' design permutations.\\n')\n",
        "                \n",
        "                # Save results\n",
        "                save, name = 0, 0\n",
        "                while save == 0:\n",
        "                    save = input('Save results as CSV file? ')\n",
        "                    if save.lower() == 'exit': sys.exit(0)\n",
        "                    save = save.lower()\n",
        "                    if save not in {'yes','y','no','n'}:\n",
        "                        save = 0\n",
        "                        print(\"Invalid input! Type 'yes' or 'no'\\n\")\n",
        "                    elif save in {'yes','y'}:\n",
        "                        while name == 0:\n",
        "                            name = input('\\nAlphanumeric filename: ')\n",
        "                            if name.lower() == 'exit': sys.exit(0)\n",
        "                            if name.isalnum():\n",
        "                                # Get rid of colorama markers\n",
        "                                for i in range(0, len(design_list[7])):\n",
        "                                    design_list[7][i] = design_list[7][i].replace('\\x1b[41m','')\n",
        "                                    design_list[7][i] = design_list[7][i].replace('\\x1b[44m','')\n",
        "                                    design_list[7][i] = design_list[7][i].replace('\\x1b[0m','')\n",
        "                                # Create a new dataframe without highlights\n",
        "                                if design == 'a': # TALENs\n",
        "                                    if TALENfor == 'a': # All targets\n",
        "                                        save_df = pd.DataFrame({'TGT':design_list[0],\n",
        "                                                                'COORD':design_list[1],\n",
        "                                                                'DES':design_list[3],\n",
        "                                                                'POTENTIAL LOSS':RE_list[0],\n",
        "                                                                'LEFT SEQUENCE':design_list[5],\n",
        "                                                                'SPACER SEQUENCE':design_list[7],\n",
        "                                                                'RIGHT SEQUENCE':design_list[6]})\n",
        "                                    elif TALENfor == 'b': # Specific site\n",
        "                                        save_df = pd.DataFrame({'DES':design_list[3],\n",
        "                                                                'POTENTIAL LOSS':RE_list[0],\n",
        "                                                                'LEFT SEQUENCE':design_list[5],\n",
        "                                                                'SPACER SEQUENCE':design_list[7],\n",
        "                                                                'RIGHT SEQUENCE':design_list[6]})\n",
        "                                else: # DdCBEs\n",
        "                                    save_df = pd.DataFrame({'TGT':design_list[0],\n",
        "                                                            'COORD':design_list[1],\n",
        "                                                            'DES':design_list[3],\n",
        "                                                            'LOSS':RE_list[0],\n",
        "                                                            'GAIN':RE_list[1],\n",
        "                                                            'OFF':design_list[4],\n",
        "                                                            'LEFT SEQUENCE':design_list[5],\n",
        "                                                            'SPACER SEQUENCE':design_list[7],\n",
        "                                                            'RIGHT SEQUENCE':design_list[6]})\n",
        "                                save_df.index += 1 # Start index at 1\n",
        "                                save_df.to_csv(name + '.csv') # Save dataframe\n",
        "                            else:\n",
        "                                name = 0\n",
        "                                print('Invalid input! File name must be alphanumeric.')\n",
        "\n",
        "                # After everything worked, ask for new sequence\n",
        "                newSequenceX = 0\n",
        "                while newSequenceX == 0:\n",
        "                    newSequenceX = input('\\nNew sequence? ')\n",
        "                    if newSequenceX.lower() == 'exit': sys.exit(0)\n",
        "                    newSequenceX = newSequenceX.lower()\n",
        "                    if newSequenceX not in {'yes','y','no','n'}:\n",
        "                        newSequenceX = 0\n",
        "                        print(\"Invalid input! Type 'yes' or 'no'\")\n",
        "                    elif newSequenceX in {'no','n'}: newSequence = False\n",
        "\n",
        "        elif not proceed: print('\\nThere are no target sites!') # If there are no targetable motifs\n",
        "    \n",
        "    # New analysis?\n",
        "    newAnalysis = 0\n",
        "    while newAnalysis == 0:\n",
        "        newAnalysis = input('\\nNew analysis? ')\n",
        "        if newAnalysis.lower() == 'exit': sys.exit(0)\n",
        "        if newAnalysis.lower() not in {'yes','y','no','n'}:\n",
        "            newAnalysis = 0\n",
        "            print(\"Invalid input! Type 'yes' or 'no'\")\n",
        "        elif newAnalysis.lower() in {'no','n'}: input('\\nBye :-)')\n",
        "        elif newAnalysis.lower() in {'yes','y'}: newAnalysis = 'yes'"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Assembly\n",
        "\n",
        "In this section, you will determine the RVDs of the TALE domains of your choosing.\n",
        "\n"
      ],
      "metadata": {
        "id": "iceHOLpqCD6G"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "v-SEoQCiSbA3",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 4. Determine the corresponding RVDs for FusX-based assembly\n",
        "\n",
        "#@markdown Please execute this cell by pressing the _Play_ button on the left. \n",
        "\n",
        "# TALE RVD SEQUENCES\n",
        "\n",
        "# Libraries\n",
        "RVDs_list, FusXB, FusB2 = ['NI','HD','NN','NG'], [], [] # A, C, G, T\n",
        "for i in range(0,4):\n",
        "    for j in range(0,4):\n",
        "        FusB2.append(RVDs_list[i] + ' ' + RVDs_list[j]) # FusB2\n",
        "        for k in range(0,4):\n",
        "            FusXB.append(RVDs_list[i] + ' ' + RVDs_list[j] + ' ' + RVDs_list[k]) # FusX1-4/B3\n",
        "\n",
        "# Number of targets\n",
        "n = 0\n",
        "while n == 0:\n",
        "    n = input('\\nNumber of targets: ')\n",
        "    if n.lower() == 'exit': sys.exit(0)\n",
        "    try:\n",
        "        n = int(n)\n",
        "        if n < 0:\n",
        "            n = 0\n",
        "            print('Invalid input! Must be a positive integer.')\n",
        "    except:\n",
        "        n = 0\n",
        "        print('Invalid input! Must be a positive integer.')\n",
        "        \n",
        "# Target sequences and corresponding RVDs\n",
        "pFusX1, pFusX2, pFusX3, pFusX4, pFusB2, pFusB3, pLR = [],[],[],[],[],[],[]\n",
        "target_list, motifs_list = [],[] # Intermediate lists\n",
        "allFusX = True # Auxiliar boolean variable\n",
        "for i in range(0, n):    \n",
        "    # Define target sequence\n",
        "    target = 0\n",
        "    while target == 0:\n",
        "        target = input('\\nTarget sequence ' + str(i+1) + ': ')\n",
        "        if target.lower() == 'exit': sys.exit(0)\n",
        "        if target == '':\n",
        "            target = 0\n",
        "            print('Invalid input! Must be a DNA sequence.')\n",
        "        else:\n",
        "            targetX = target\n",
        "            target = target.lower()\n",
        "            j = 0\n",
        "            while j in range(0, len(targetX)):\n",
        "                if target[j] not in {'a','c','g','t'}:\n",
        "                    j = len(target)\n",
        "                    target = 0\n",
        "                    print('Invalid input! Must be a DNA sequence.')\n",
        "                else: j += 1\n",
        "    target_list.append(target)\n",
        "    \n",
        "    # Auxiliar boolean statement to avoid displaying errors\n",
        "    if not 15 <= len(target) <= 16:\n",
        "        print('\\nWarning! Target length is not FusX-compatible.')\n",
        "        allFusX = False\n",
        "                    \n",
        "    # Determine RVDs\n",
        "    specificMotif = ''\n",
        "    for i in range(0, len(target)):\n",
        "        if i == len(target) - 1:\n",
        "            if target[i] == 'a': specificMotif = specificMotif + RVDs_list[0]\n",
        "            elif target[i] == 'c': specificMotif = specificMotif + RVDs_list[1]\n",
        "            elif target[i] == 'g': specificMotif = specificMotif + RVDs_list[2]\n",
        "            elif target[i] == 't': specificMotif = specificMotif + RVDs_list[3]         \n",
        "        else:\n",
        "            if target[i] == 'a': specificMotif = specificMotif + RVDs_list[0] + ' '\n",
        "            elif target[i] == 'c': specificMotif = specificMotif + RVDs_list[1] + ' '\n",
        "            elif target[i] == 'g': specificMotif = specificMotif + RVDs_list[2] + ' '\n",
        "            elif target[i] == 't': specificMotif = specificMotif + RVDs_list[3] + ' ' \n",
        "    motifs_list.append(specificMotif)\n",
        "    \n",
        "    # Determine plasmids\n",
        "    if 15 <= len(target) <= 16:\n",
        "        for i in range(0,len(FusXB)):\n",
        "            if FusXB[i] == specificMotif[0:8]:\n",
        "                pFusX1.append(i+1)\n",
        "            if FusXB[i] == specificMotif[9:17]:\n",
        "                pFusX2.append(i+1)\n",
        "            if FusXB[i] == specificMotif[18:26]:\n",
        "                pFusX3.append(i+1)\n",
        "            if FusXB[i] == specificMotif[27:35]:\n",
        "                pFusX4.append(i+1)\n",
        "        if len(target) == 15:\n",
        "            for i in range(0,len(FusB2)):\n",
        "                if FusB2[i] == specificMotif[36:41]:\n",
        "                    pFusB2.append(i+1)\n",
        "                    pFusB3.append(' ')\n",
        "            pLR.append(specificMotif[42:])\n",
        "        else:\n",
        "            for i in range(0,len(FusXB)):\n",
        "                if FusXB[i] == specificMotif[36:44]:\n",
        "                    pFusB2.append(' ')\n",
        "                    pFusB3.append(i+1)\n",
        "            pLR.append(specificMotif[45:])  \n",
        "\n",
        "# Display results\n",
        "if allFusX == True: # Include plasmids only if all target sequences are FusX-compatible\n",
        "    df = pd.DataFrame({'Target sequence': target_list,\n",
        "                       'Corresponding RVDs': motifs_list,\n",
        "                       'X1':pFusX1,\n",
        "                       'X2':pFusX2,\n",
        "                       'X3':pFusX3,\n",
        "                       'X4':pFusX4,\n",
        "                       'B2':pFusB2,\n",
        "                       'B3':pFusB3,\n",
        "                       'LR':pLR})\n",
        "else: df = pd.DataFrame({'Target sequence': target_list, 'Corresponding RVDs': motifs_list})\n",
        "print('\\n' + tabulate(df, showindex=False, headers=df.columns))\n",
        "\n",
        "# Save results\n",
        "save, name = 0, 0\n",
        "while save == 0:\n",
        "    save = input('\\nSave results as CSV file? ')\n",
        "    if save.lower() == 'exit': sys.exit(0)\n",
        "    save = save.lower()\n",
        "    if save not in {'yes','y','no','n'}:\n",
        "        save = 0\n",
        "        print(\"Invalid input! Type 'yes' or 'no'\\n\")\n",
        "    elif save in {'yes','y'}:\n",
        "        while name == 0:\n",
        "            name = input('\\nAlphanumeric filename: ')\n",
        "            if name.lower() == 'exit': sys.exit(0)\n",
        "            if name.isalnum():\n",
        "                df.index += 1 # Start index at 1\n",
        "                df.to_csv(name + '.csv') # Save dataframe\n",
        "            else:\n",
        "                name = 0\n",
        "                print('Invalid input! File name must be alphanumeric.')\n",
        "                \n",
        "# Exit\n",
        "boolX = False\n",
        "while boolX == False:\n",
        "    boolX = True\n",
        "    input('\\nBye :-)')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# FAQ & Troubleshooting\n",
        "\n",
        "* How long will this take?\n",
        "  * Downloading and installing the third-party software can take up to a few minutes.\n",
        "  * Defining the libraries and functions can take up to a few minutes.\n",
        "  * Running TALE Writer can take up to a few minutes. \n",
        "* My Colab no longer seems to be doing anything, what should I do?\n",
        "  * If nothing happens or if you receive an error message, try restarting your Colab runtime via *Runtime > Restart runtime*.\n",
        "  * If this doesn't help, try resetting your Colab runtime via *Runtime > Factory reset runtime*.\n",
        "*   What is a Colab?\n",
        "  *   See the [Colab FAQ](https://research.google.com/colaboratory/faq.html).\n",
        "* I received an error \"Colab CPU runtime not supported\" or \"No GPU/TPU found\", what do I do?\n",
        "  * Colab CPU runtime is not supported. Try changing your runtime via *Runtime > Change runtime type > Hardware accelerator > GPU*.\n",
        "  * [Colab Pro](https://colab.research.google.com/signup) offers priority access to GPUs.\n",
        "* I received an error \"ModuleNotFoundError: No module named...\", even though I ran the cell that imports it, what do I do?\n",
        "  * Colab notebooks on the free tier time out after a certain amount of time. See the [Colab FAQ](https://research.google.com/colaboratory/faq.html). Try rerunning the whole notebook from the beginning.\n",
        "* How should I share feedback and bug reports?\n",
        "  * Please share any feedback and bug reports as an [issue](https://github.com/srcastillo/TALE-Writer/issues) on Github."
      ],
      "metadata": {
        "id": "uxwW6UukhCPw"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# License and Disclaimer\n",
        "\n",
        "This is not an officially-supported Google product.\n",
        "\n",
        "This Colab notebook and other information provided is for computer-aided design of TALENs and FusXTBEs only, caution should be exercised in its use. It is porivded 'as-is' without any warranty of any kind, whether expressed or implied. Information is not inteded to be a substitute for professional medical advice, diagnosis, or treatment, and does not constitute medial or other professional advice.\n",
        "\n",
        "## TALE Writer Code License\n",
        "\n",
        "Licensed under the GNU General Public License, Version 3.0 (the \"License\"). You may obtain a copy of the License at https://www.gnu.org/licenses/gpl-3.0.en.html.\n",
        "\n",
        "## Third-party software\n",
        "\n",
        "Use of the third-party software, libraries or code referred to in the [Acknowledgements](https://github.com/srcastillo/TALE-Writer/blob/main/README.md#acknowledgements) section in the TALE Writer README may be governed by separate terms and conditions or license provisions. Your use of the third-party software, libraries or code is subject to any such terms and you should check that you can comply with any applicable restrictions or terms and conditions before use.\n",
        "\n"
      ],
      "metadata": {
        "id": "eQx-cJrUhiOa"
      }
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.3"
    },
    "colab": {
      "name": "Copy of TALE_Writer_Colab.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}